packages/scoped-elements/html-element.js

```diff
@@ -5,20 +5,21 @@ import { dedupeMixin } from '@open-wc/dedupe-mixin';
  * @typedef {import('./types.js').ScopedElementsMap} ScopedElementsMap
  */
 
-const version = '3.0.0';
-const versions = window.scopedElementsVersions || (window.scopedElementsVersions = []);
-if (!versions.includes(version)) {
-  versions.push(version);
-}
-
 /**
  * @template {import('./types.js').Constructor<HTMLElement>} T
  * @param {T} superclass
  * @return {T & import('./types.js').Constructor<ScopedElementsHost>}
  */
-const ScopedElementsMixinImplementation = superclass =>
+const ScopedElementsMixinImplementation = superclass => {
+  const version = '3.0.0';
+
+  const versions = window.scopedElementsVersions || (window.scopedElementsVersions = []);
```

Comment:
Also does it actually matter for the side-effect analysis to move it inside the closure? If the parent module imports and calls `ScopedElementsMixin()`, the effect would be the same, right?

Reply:
> Is it not enough to mark it as side-effects: false?

I've been trying to learn more about dead code elimination to figure out answers to questions like this. Right now I don't really know. I've been kind of assuming that _all_ code that executes when a module is imported is side-effects but maybe not? Making changes to `window` _does_ feel like side-effects to me though.

Reply:
> Also does it actually matter for the side-effect analysis to move it inside the closure? If the parent module imports and calls `ScopedElementsMixin()`, the effect would be the same, right?

I think a bundler can eliminate everything other than the imported function during dead code elimination if the file has no side-effects. When you do have top level code, you're gonna need to import the whole file as you need to execute the top-level code.

Reply:
Its definitely a sideeffect, but that doesnt mean it shouldnt be treeshaken :p In this case it should be safe, because it's only a write, nothing depends on reading the variable. Hence the question if its not enough to mark the package as sideEffects: false?

Reply:
I figured out a way to check. [The Rollup REPL default tree shake demo has a example that's really similar to this code 😄 ](https://rollupjs.org/repl/?version=4.9.6&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMFRSRUUtU0hBS0lORyU1Q25pbXBvcnQlMjAlN0IlMjBjdWJlJTIwJTdEJTIwZnJvbSUyMCcuJTJGbWF0aHMuanMnJTNCJTVDbiU1Q25jb25zb2xlLmxvZyhjdWJlKDUpKSUzQiUyMCUyRiUyRiUyMDEyNSUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyJTJGJTJGJTIwbWF0aHMuanMlNUNuJTVDbiUyRiUyRiUyMFRoaXMlMjBmdW5jdGlvbiUyMGlzbid0JTIwdXNlZCUyMGFueXdoZXJlJTJDJTIwc28lNUNuJTJGJTJGJTIwUm9sbHVwJTIwZXhjbHVkZXMlMjBpdCUyMGZyb20lMjB0aGUlMjBidW5kbGUuLi4lNUNuZXhwb3J0JTIwY29uc3QlMjBzcXVhcmUlMjAlM0QlMjB4JTIwJTNEJTNFJTIweCUyMColMjB4JTNCJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwZnVuY3Rpb24lMjBnZXRzJTIwaW5jbHVkZWQlNUNuJTJGJTJGJTIwcmV3cml0ZSUyMHRoaXMlMjBhcyUyMCU2MHNxdWFyZSh4KSUyMColMjB4JTYwJTVDbiUyRiUyRiUyMGFuZCUyMHNlZSUyMHdoYXQlMjBoYXBwZW5zISU1Q25leHBvcnQlMjBjb25zdCUyMGN1YmUlMjAlM0QlMjB4JTIwJTNEJTNFJTIweCUyMColMjB4JTIwKiUyMHglM0IlNUNuJTVDbiUyRiUyRiUyMFRoaXMlMjAlNUMlMjJzaWRlJTIwZWZmZWN0JTVDJTIyJTIwY3JlYXRlcyUyMGElMjBnbG9iYWwlNUNuJTJGJTJGJTIwdmFyaWFibGUlMjBhbmQlMjB3aWxsJTIwbm90JTIwYmUlMjByZW1vdmVkLiU1Q253aW5kb3cuZWZmZWN0MSUyMCUzRCUyMCdjcmVhdGVkJyUzQiU1Q24lNUNuY29uc3QlMjBpbmNsdWRlRWZmZWN0JTIwJTNEJTIwZmFsc2UlM0IlNUNuaWYlMjAoaW5jbHVkZUVmZmVjdCklMjAlN0IlNUNuJTVDdCUyRiUyRiUyME9uJTIwdGhlJTIwb3RoZXIlMjBoYW5kJTJDJTIwdGhpcyUyMGlzJTIwbmV2ZXIlNUNuJTVDdCUyRiUyRiUyMGV4ZWN1dGVkJTIwYW5kJTIwdGh1cyUyMHJlbW92ZWQuJTVDbiU1Q3R3aW5kb3cuZWZmZWN0MSUyMCUzRCUyMCdub3QlMjBjcmVhdGVkJyUzQiU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIybWF0aHMuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE).

The comments indicate that it is a side-effect but that it won't be removed so it looks like this is all good.

> In this case it should be safe, because it's only a write, nothing depends on reading the variable.

Where do you know this from? I've been trying to look up info on how side-effects and tree shaking heuristics work and I can't find a definite guide to what can be tree shaken and what can't.

Reply:
> > Also does it actually matter for the side-effect analysis to move it inside the closure? If the parent module imports and calls `ScopedElementsMixin()`, the effect would be the same, right?
> 
> I think a bundler can eliminate everything other than the imported function during dead code elimination if the file has no side-effects. When you do have top level code, you're gonna need to import the whole file as you need to execute the top-level code.

This turned out to be not true lol

Reply:
I think I just don't fully understand side-effects and treeshaking. Rollup seems to handle everything I throw at it. Does it then _need_ this flag in the package.json? Couldn't the bundler just figure it out by analysing the code?

Reply:
Maybe it's different because of `@rollup/plugin-node-resolve`.

Reply:
> Couldn't the bundler just figure it out by analysing the code

I would have assumed that to be the case yeah

Reply:
> Where do you know this from? I've been trying to look up info on how side-effects and tree shaking heuristics work and I can't find a definite guide to what can be tree shaken and what can't.

Past experiences and looking at build output I guess. As far as I'm aware there's not some great writeup on side-effects (would be nice to have a resource like that), but I learn by seeing/doing, so I often make minimal repros and try things out to see how different things lead to different build output.